### DISCUSSION TITLE
post/ipv4-tcp-hole-punching/

### DISCUSSION THREAD

**Discussion by giscus** [lyc8503/lyc8503#13] [Announcements]
# post/ipv4-tcp-hole-punching/

是的, 你没有看错, 我也没有写错, 家庭宽带 IPv4 的 TCP 也是可以打洞的… 如果是我火星请原谅我. Anyway, 分享如下. 为了在浪急风高, 晦暗莫测的公网建立起 笔记本 -> HomeLab 快速安全的通道, 我已经研究过很多很多种方法了… 简要列举如下:  申请公网 IPv4 放弃: 我是移动用户, 没 IPv4  使用公网 IPv6 + DDNS 目前正作为备用方案,

https://blog.lyc8503.site/post/ipv4-tcp-hole-punching/

**Comment by ZHH-start** [comment]
看起来很不错。不过我担心的是安全性。我前几天在Bing上搜索“qb”关键词，意外地搜到至少四五个内网qb客户端通过DDNS+域名到外网，而且甚至连密码都没设置。

**Reply by lyc8503** [reply]
暴露到公网还是推荐使用 VPN, 这篇只是在讲暴露端口的方法, 安全问题还是需要自己考虑...

**Comment by ZHH-start** [comment]
你说的很有道理，我正希望用VPN。不过其实这些都需要花时间维护，如果不是那么紧急的需求，似乎大部分都可以用SSD+硬盘盒解决。啊……我觉得其实我好像也没有那么需要外网访问家里、不管如何，先收藏吧&nbsp;


哈哈公主
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2023年10月30日(星期一) 下午2:23
收件人: "lyc8503/lyc8503";
抄送: "Comment";
主题: Re: [lyc8503/lyc8503] post/ipv4-tcp-hole-punching/ (Discussion #13)





 
暴露到公网还是推荐使用 VPN, 这篇只是在讲暴露端口的方法, 安全问题还是需要自己考虑...
 
&mdash;

**Comment by liudongmiao** [comment]
可以考虑 tinc vpn: https://www.tinc-vpn.org/

NAT traversal: As long as one node in the VPN allows incoming connections on a public IP address (even if it is a dynamic IP address), tinc will be able to do NAT traversal, allowing direct communication between peers.

**Reply by ZHH-start** [reply]
感谢。我了解一下，看起来这个VPN不要求公网？

**Reply by ZHH-start** [reply]
好吧，这似乎是一个专用虚拟网络。需要一个服务器当作核心。听起来挺好的，有时间我会部署一下试试的，谢谢你的分享

**Reply by lyc8503** [reply]
这个和我提到的 Tailscale, Zerotier 算是比较相近的方法, TCP 打洞的话可以暴露一个直接公网直接可访问的端口, 无需特殊的 Agent 直接连接可能方便一点~ 不过 P2P VPN 也是不错的一类解决方案.

**Reply by ZHH-start** [reply]
收到

**Comment by 1c383r9** [comment]
我用的 ip4p 方案，虽然我觉得 srv 记录就挺合适，但是 mihomo 内核支持 ip4p 而不支持 srv, 这样回家太方便了，香。
